File: solutions/382/main.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

import random 
random.seed(0)
from typing import Optional

class Solution:
    def __init__(self, head: Optional[ListNode]):
        self.head = head
    def getRandom(self) -> int:
        result = self.head.val
        cur = self.head
        count = 0
        while cur:
            if random.randint(0, count) == 0:
                result = cur.val
            count += 1
            cur = cur.next
        return result

if __name__ == "__main__": 
    head = ListNode(0)
    head.next = ListNode(1)

    obj = Solution(head)
    param_1 = obj.getRandom()
    print(param_1)

File: solutions/382/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sun Jan 16 09:35:20 2022
> Description:   
 ************************************************************************/
#include <iostream> 
#include <memory> 
#include <stdlib.h>
#include <gtest/gtest.h> 

using namespace std; 

// Definition for singly-linked list.
template <typename T> 
struct ListNode {
    T val;
    shared_ptr<ListNode<T>> next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(T x) : val(x), next(nullptr) {}
    ListNode(T x, shared_ptr<ListNode<T>> next) : val(x), next(next) {}
};

template <typename T> 
class Solution {
private:
    shared_ptr<ListNode<T>> _head;
public:
    Solution(shared_ptr<ListNode<T>> head) {
        _head = head;
        srand(0);
    }
    int getRandom() {
        shared_ptr<ListNode<T>> cur = _head;
        int result = cur->val;
        int count = 0;
        while (cur != nullptr) {
            count++;
            if (rand() % count == 1) result = cur->val;
            cur = cur->next;
        }
        return result;
    }
};

TEST(Test382, SimpleTest) {
    auto head = make_shared<ListNode<int>>(0);
    head->next = make_shared<ListNode<int>>(1);

    Solution<int> s(head); 
    cout << "random number: " << s.getRandom() << endl;
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

